File: Python/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-Histogram.py
'''
Largest Rectangle in Histogram - Stack-Based Solutions

Problem Statement:
Given an array of heights representing the histogram's bar heights where the width of each bar is 1,
return the area of the largest rectangle in the histogram.
'''

# 1. Monotonic Stack - O(N) Time Complexity
def largestRectangleArea_stack(heights):
    """
    Calculates the largest rectangle area in a histogram using a monotonic stack.
    A monotonic stack maintains a strictly increasing or decreasing order of elements.
    This approach is efficient because it processes each bar only a constant number of times.

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    stack = []  # Stores (index, height) of bars. The stack maintains increasing heights.
    max_area = 0
    heights.append(0)  # Sentinel value to ensure all bars are processed from the stack.

    for i, h in enumerate(heights):
        # Iterate through the heights and their indices.
        while stack and stack[-1][1] > h:
            # While the stack is not empty and the current height is less than the height of the bar at the top of the stack...
            _, height = stack.pop()  # ...pop the bar from the stack.
            width = i if not stack else i - stack[-1][0] - 1
            # Calculate the width of the rectangle.
            # If the stack is empty, the width extends to the beginning of the histogram.
            # Otherwise, the width is the distance between the current bar and the previous bar in the stack.
            max_area = max(max_area, height * width)  # Update the maximum area.
        stack.append((i, h))  # Push the current bar's index and height onto the stack.

    return max_area  # Return the maximum area.

# 2. Divide and Conquer - O(N log N) Time Complexity
def largestRectangleArea_divide_conquer(heights):
    """
    Calculates the largest rectangle area in a histogram using a divide and conquer approach.
    This approach recursively divides the histogram into subproblems and combines the solutions.
    The time complexity is O(N log N) on average, but can be O(N^2) in the worst case (for sorted input).

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    def compute_area(left, right):
        """
        Recursively computes the largest rectangle area within the given range.

        Args:
            left: The left index of the range.
            right: The right index of the range.

        Returns:
            The area of the largest rectangle within the range.
        """
        if left > right:
            return 0  # Base case: empty range.
        min_index = min(range(left, right + 1), key=lambda i: heights[i])
        # Find the index of the minimum height bar in the range.
        return max(
            heights[min_index] * (right - left + 1),  # Area of the rectangle with the minimum height bar.
            compute_area(left, min_index - 1),  # Recursively find the largest area in the left subrange.
            compute_area(min_index + 1, right)  # Recursively find the largest area in the right subrange.
        )

    return compute_area(0, len(heights) - 1)  # Start the recursion with the entire range.

# 3. Two Pass with Left & Right Boundaries - O(N) Time Complexity
def largestRectangleArea_two_pass(heights):
    """
    Calculates the largest rectangle area in a histogram using two passes to find left and right boundaries.
    This approach precomputes the left and right boundaries for each bar, which define the maximum width
    of a rectangle with that bar as the minimum height.

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    n = len(heights)
    left, right = [0] * n, [0] * n  # Arrays to store the left and right boundaries for each bar.
    stack = []

    # Calculate the left boundaries.
    for i in range(n):
        while stack and heights[stack[-1]] >= heights[i]:
            stack.pop()
        left[i] = stack[-1] if stack else -1  # The left boundary is the index of the first bar to the left with a height less than heights[i]
        stack.append(i)

    stack.clear()  # Clear the stack for the next pass.

    # Calculate the right boundaries.
    for i in range(n - 1, -1, -1):
        while stack and heights[stack[-1]] >= heights[i]:
            stack.pop()
        right[i] = stack[-1] if stack else n # The right boundary is the index of the first bar to the right with a height less than heights[i]
        stack.append(i)

    # Calculate the maximum area.
    return max(heights[i] * (right[i] - left[i] - 1) for i in range(n))

# 4. Brute Force - O(N^2) Time Complexity
def largestRectangleArea_brute_force(heights):
    """
    Calculates the largest rectangle area in a histogram using a brute-force approach.
    This approach iterates through all possible rectangles and calculates their areas,
    resulting in a time complexity of O(N^2), which is inefficient for large datasets.

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    max_area = 0
    n = len(heights)

    for i in range(n):
        min_height = heights[i]  # Start with the current bar's height as the minimum.
        for j in range(i, n):
            min_height = min(min_height, heights[j])  # Find the minimum height in the current range.
            max_area = max(max_area, min_height * (j - i + 1))  # Calculate the area and update the maximum area.

    return max_area

# 5. Dynamic Programming - O(N) Time Complexity
def largestRectangleArea_dp(heights):
    """
    Calculates the largest rectangle area in a histogram using dynamic programming.
    This approach is similar to the two-pass approach, but it uses dynamic programming
    to efficiently calculate the left and right boundaries.

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    n = len(heights)
    left, right = [0] * n, [0] * n

    # Calculate the left boundaries using dynamic programming.
    for i in range(n):
        p = i - 1
        while p >= 0 and heights[p] >= heights[i]:
            p = left[p] - 1  # Use previously computed left boundaries.
        left[i] = p + 1

    # Calculate the right boundaries using dynamic programming.
    for i in range(n - 1, -1, -1):
        p = i + 1
        while p < n and heights[p] >= heights[i]:
            p = right[p] + 1  # Use previously computed right boundaries.
        right[i] = p - 1

    # Calculate the maximum area.
    return max(heights[i] * (right[i] - left[i] + 1) for i in range(n))

# Test Cases
heights = [2, 1, 5, 6, 2, 3]
print("Monotonic Stack:", largestRectangleArea_stack(heights))
print("Divide and Conquer:", largestRectangleArea_divide_conquer(heights))
print("Two Pass:", largestRectangleArea_two_pass(heights))
print("Brute Force:", largestRectangleArea_brute_force(heights))
print("Dynamic Programming:", largestRectangleArea_dp(heights))

File: JavaScript/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-Histogram.js
/*
    Largest Rectangle in Histogram: 5 Approaches in JavaScript

    Problem:
    Given an array of integers heights representing the histogram's bar heights
    where the width of each bar is 1, find the area of the largest rectangle in the histogram.

    Constraints:
    - 1 <= heights.length <= 10^5
    - 0 <= heights[i] <= 10^4
*/

/**
 * Approach 1: Brute Force
 * For each bar, expand left and right to find the maximum width rectangle.
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 */
function largestRectangleAreaBruteForce(heights) {
    let maxArea = 0;
    const n = heights.length;

    for (let i = 0; i < n; i++) {
        let width = 1;
        let left = i - 1;
        let right = i + 1;

        // Expand left
        while (left >= 0 && heights[left] >= heights[i]) {
            width++;
            left--;
        }

        // Expand right
        while (right < n && heights[right] >= heights[i]) {
            width++;
            right++;
        }

        maxArea = Math.max(maxArea, heights[i] * width);
    }

    return maxArea;
}

/**
 * Approach 2: Divide and Conquer
 * Recursively divide the histogram and find the largest rectangle in each half.
 * The largest rectangle can also span across the two halves.
 * Time Complexity: O(n log n) in average case, O(n^2) in worst case (if the array is sorted)
 * Space Complexity: O(n) due to recursion stack
 */
function largestRectangleAreaDivideAndConquer(heights, start, end) {
    if (start > end) {
        return 0;
    }

    if (start === end) {
        return heights[start];
    }

    let minIndex = start;
    for (let i = start + 1; i <= end; i++) {
        if (heights[i] < heights[minIndex]) {
            minIndex = i;
        }
    }

    const leftArea = largestRectangleAreaDivideAndConquer(heights, start, minIndex - 1);
    const rightArea = largestRectangleAreaDivideAndConquer(heights, minIndex + 1, end);
    const midArea = heights[minIndex] * (end - start + 1);

    return Math.max(leftArea, rightArea, midArea);
}

function largestRectangleAreaDivideAndConquerHelper(heights) {
    return largestRectangleAreaDivideAndConquer(heights, 0, heights.length - 1);
}


/**
 * Approach 3: Stack
 * Use a stack to keep track of increasing bar indices.
 * When a smaller bar is encountered, calculate the area with the previous bars in the stack.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaStack(heights) {
    let maxArea = 0;
    const n = heights.length;
    const stack = [];
    let i = 0;

    while (i < n || stack.length > 0) { // Corrected loop condition
        if (i < n && (stack.length === 0 || heights[i] >= heights[stack[stack.length - 1]])) {
            stack.push(i);
            i++;
        } else {
            const top = stack.pop();
            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, heights[top] * width);
        }
    }
    return maxArea;
}

/**
 * Approach 4: Stack (Optimized)
 * Optimized version of the stack approach.  Append 0 to the heights array to ensure
 * all elements in the stack are processed.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaStackOptimized(heights) {
    let maxArea = 0;
    const n = heights.length;
    const stack = [];
    const modifiedHeights = [...heights, 0]; // Append 0 to handle remaining bars in stack

    for (let i = 0; i < modifiedHeights.length; i++) {
        while (stack.length > 0 && modifiedHeights[i] < modifiedHeights[stack[stack.length - 1]]) {
            const top = stack.pop();
            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, modifiedHeights[top] * width);
        }
        stack.push(i);
    }

    return maxArea;
}

/**
 * Approach 5: Using Min-Left and Min-Right Arrays
 * Calculate the leftmost and rightmost boundaries for each bar that are greater or equal to the current bar.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function largestRectangleAreaMinLeftRight(heights) {
    const n = heights.length;
    const left = new Array(n);
    const right = new Array(n);
    let maxArea = 0;

    // Calculate left boundaries
    left[0] = 0;
    for (let i = 1; i < n; i++) {
        let p = i - 1;
        while (p >= 0 && heights[p] >= heights[i]) {
            p = left[p] - 1;
        }
        left[i] = Math.max(0, p + 1); // Ensure left[i] is not negative
    }

    // Calculate right boundaries
    right[n - 1] = n - 1;
    for (let i = n - 2; i >= 0; i--) {
        let p = i + 1;
        while (p < n && heights[p] >= heights[i]) {
            p = right[p] + 1;
        }
        right[i] = Math.min(n - 1, p - 1); // Ensure right[i] is not out of bounds.
    }

    // Calculate max area
    for (let i = 0; i < n; i++) {
        maxArea = Math.max(maxArea, heights[i] * (right[i] - left[i] + 1));
    }
    return maxArea;
}



// Example Usage:
const heights = [2, 1, 5, 6, 2, 3];
console.log("Heights:", heights);

// Brute Force
console.log("Largest Rectangle Area (Brute Force):", largestRectangleAreaBruteForce(heights)); // Output: 10

// Divide and Conquer
console.log("Largest Rectangle Area (Divide and Conquer):", largestRectangleAreaDivideAndConquerHelper(heights)); // Output: 10

// Stack
console.log("Largest Rectangle Area (Stack):", largestRectangleAreaStack(heights)); // Output: 10

// Stack Optimized
console.log("Largest Rectangle Area (Stack Optimized):", largestRectangleAreaStackOptimized(heights)); // Output: 10

// Min-Left and Min-Right Arrays
console.log("Largest Rectangle Area (Min-Left/Right):", largestRectangleAreaMinLeftRight(heights)); // Output: 10
